src/main/kotlin/snyk/oss/annotator/OSSBaseAnnotator.kt

```diff
@@ -0,0 +1,144 @@
+package snyk.oss.annotator
+
+import com.intellij.lang.annotation.AnnotationBuilder
+import com.intellij.lang.annotation.AnnotationHolder
+import com.intellij.lang.annotation.ExternalAnnotator
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.progress.ProgressManager
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiFile
+import io.snyk.plugin.getSnykToolWindowPanel
+import snyk.common.intentionactions.AlwaysAvailableReplacementIntentionAction
+import snyk.oss.OssVulnerabilitiesForFile
+import snyk.oss.Vulnerability
+import snyk.oss.annotator.AnnotatorHelper.severity
+import kotlin.math.max
+
+abstract class OSSBaseAnnotator : ExternalAnnotator<PsiFile, Unit>() {
+    // overrides needed for the Annotator to invoke apply(). We don't do anything here
+    override fun collectInformation(file: PsiFile): PsiFile = file
+    override fun doAnnotate(psiFile: PsiFile?) {
+        val filePath = psiFile?.virtualFile?.path ?: return
+
+        if (AnnotatorHelper.isFileSupported(filePath)) {
+            val document = psiFile.viewProvider.document ?: return
+            ApplicationManager.getApplication().invokeAndWait {
+                FileDocumentManager.getInstance().saveDocument(document)
+            }
+        }
+    }
+
+    override fun apply(psiFile: PsiFile, annotationResult: Unit, holder: AnnotationHolder) {
+        val issues = getIssuesForFile(psiFile) ?: return
+        val remediation = issues.remediation
+
+        issues.vulnerabilities
+            .forEach { vulnerability ->
+                if (vulnerability.ignored || vulnerability.obsolete) return@forEach
+
+                val severity = severity(vulnerability)
+                val textRange = textRange(psiFile, vulnerability)
+                val annotationBuilder =
+                    holder.newAnnotation(severity, annotationMessage(vulnerability)).range(textRange)
+                val fixRange = fixRange(psiFile, vulnerability)
+                if (fixRange != TextRange.EMPTY_RANGE && remediation != null && remediation.upgrade.isNotEmpty()) {
+                    addQuickFix(psiFile, vulnerability, annotationBuilder, fixRange, remediation)
+                }
+                annotationBuilder.create()
+            }
+    }
+
+    open fun getIssuesForFile(psiFile: PsiFile): OssVulnerabilitiesForFile? {
+        val ossResult = getSnykToolWindowPanel(psiFile.project)?.currentOssResults ?: return null
+        val filePath = psiFile.virtualFile?.path ?: return null
+        !AnnotatorHelper.isFileSupported(filePath) && return null
```

Comment:
classic `if` would be more readable here

Reply:
I like it this way, but will change it.